Code/GraphMol/FMCS/MaximumCommonSubgraph.cpp

```diff
@@ -205,66 +205,51 @@ void MaximumCommonSubgraph::init() {
-    for (ROMol::ConstAtomIterator a = Targets[i].Molecule->beginAtoms();
-         a != Targets[i].Molecule->endAtoms(); a++, j++) {
-      Targets[i].Topology.addAtom((*a)->getIdx());
-    }
-    // build Target Topology ADD BONDs
-    for (ROMol::ConstBondIterator b = Targets[i].Molecule->beginBonds();
-         b != Targets[i].Molecule->endBonds(); b++) {
-      const Bond* bond = *b;
-      unsigned ii = bond->getBeginAtomIdx();
-      unsigned jj = bond->getEndAtomIdx();
-      Targets[i].Topology.addBond((*b)->getIdx(), ii, jj);
-    }
+}
 
-#ifdef FAST_SUBSTRUCT_CACHE
-    // fill out RingMatchTables
-    if (!userData  // predefined functor - compute RingMatchTable for all
-                   // targets
-        && (Parameters.BondCompareParameters.CompleteRingsOnly ||
-            Parameters.BondCompareParameters.RingMatchesRingOnly)) {
-      RingMatchTables.addTargetBondRingsIndeces(Targets[i].Molecule);
-      RingMatchTables.computeRingMatchTable(QueryMolecule, Targets[i].Molecule,
-                                            Parameters);
+void MaximumCommonSubgraph::loadMatchTables(const ROMol* targetMolecule,
+                                            MatchTable& atomMatchTable,
+                                            MatchTable& bondMatchTable) {
+  // fill out RingMatchTables to check cache Hash collision by checking match a
+  // part of Query to Query
+  // predefined functor - compute RingMatchTable for all targets
+  if (Parameters.BondCompareParameters.CompleteRingsOnly ||
+      Parameters.BondCompareParameters.RingMatchesRingOnly ||
+      Parameters.AtomCompareParameters.RingMatchesRingOnly) {
+    if (QueryMolecule == targetMolecule) {
+      RingMatchTables.init(QueryMolecule);
+      CompareFunctionsData.ringMatchTables = &RingMatchTables;
+    } else {
+      RingMatchTables.addTargetBondRingsIndeces(targetMolecule);
     }
-#endif
+    RingMatchTables.computeRingMatchTable(QueryMolecule, targetMolecule,
+                                          Parameters, CompareFunctionsData);
+  }
 
-    // fill out match tables
-    size_t nq = QueryMolecule->getNumAtoms();
-    size_t nt = (*it)->getNumAtoms();
-    Targets[i].AtomMatchTable.resize(nq, nt);
-
-    for (size_t aj = 0; aj < nt; aj++) {
-      for (size_t ai = 0; ai < nq; ai++) {
-        Targets[i].AtomMatchTable.set(
-            ai, aj,
-            Parameters.AtomTyper(Parameters.AtomCompareParameters,
-                                 *QueryMolecule, ai, *Targets[i].Molecule, aj,
-                                 Parameters.CompareFunctionsUserData));
-      }
+  // fill out match tables
+  size_t nq = QueryMolecule->getNumAtoms();
+  size_t nt = targetMolecule->getNumAtoms();
+
+  atomMatchTable.resize(nq, nt);
+  for (size_t aj = 0; aj < nt; aj++) {
+    for (size_t ai = 0; ai < nq; ai++) {
+      atomMatchTable.set(
+          ai, aj,
+          Parameters.AtomTyper(Parameters.AtomCompareParameters, *QueryMolecule,
```

Comment:
I believe there's an opportunity to improve performance but I was already changing enough here that I didn't do it. Every time we check a different input molecule as the query molecule, we rerun these calculations. But we could cache the results based on the atom and bond pointers. I would be happy to make that change.